File: src/stats/sg-baseline.ts
// Which handicap cohort the strokes-gained-lite attribution is measured
// against (`docs/proposals/strokes-gained-lite.md`, "Handicap cohorts").
//
// The tables themselves live in `src/round/stat-measures.ts`, which has no
// runtime imports because it is ported line for line to Swift. THIS file is the
// app-side half: the stored choice, the words a player reads, and the one
// function that turns a choice plus a handicap into a cohort. Keep the pure
// math there and the product decisions here.
//
// Twin of `ios/TapScore/Features/Stats/SgBaseline.swift`.

import { defaultStorage, deviceStore, type DeviceStorage } from '../device-store';
import {
    cohortForHandicap,
    expectedOnParSeventyTwo,
    SG_BASELINES_V1,
    SG_COHORTS,
    type SgCohort,
} from '../round/stat-measures';

/**
 * What the player picked in Filters: a fixed tier, or `auto` — "match my
 * handicap", which re-resolves as their index moves.
 */
export type SgBaselineChoice = 'auto' | SgCohort;

/**
 * Picker order: the auto mode first, then the four tiers easiest to hardest.
 *
 * `auto` leads because it is the answer for almost everybody — the tier list
 * below it is for the player who wants to see their game against a standard
 * that is not their own.
 */
export const SG_BASELINE_CHOICES: readonly SgBaselineChoice[] = ['auto', ...SG_COHORTS];

/**
 * The default: follow the handicap. A player who never opens Filters is
 * measured against the tier that fits them, and a player with no handicap on
 * file is measured against `hcp12` — exactly what every reading in the app
 * meant before cohorts existed.
 */
export const FALLBACK_SG_CHOICE: SgBaselineChoice = 'auto';

/**
 * A cohort's name, in the words both clients use. No "scr"/"hcp5" shorthand
 * reaches a screen, and no symbol stands in for a word.
 */
export function cohortLabel(cohort: SgCohort): string {
    switch (cohort) {
        case 'scratch':
            return 'Scratch';
        case 'hcp5':
            return '5 handicap';
        case 'hcp12':
            return '12 handicap';
        case 'hcp20':
            return '20+ handicap';
    }
}

/** The picker row's label — the tiers by name, plus the auto mode. */
export function choiceLabel(choice: SgBaselineChoice): string {
    return choice === 'auto' ? 'Match my handicap' : cohortLabel(choice);
}

/**
 * The muted line under the picker: what the choice MEANS for this reader.
 *
 * A tier answers in the only unit a golfer can check themselves against — a
 * score on a par 72, DERIVED from that tier's own table, so it cannot drift out
 * of step with the numbers it describes. The auto row instead says which tier
 * the reader's own index lands on, so choosing between the rows never requires
 * knowing where the boundaries are.
 */
export function choiceHint(choice: SgBaselineChoice, handicapIndex: number | null): string {
    if (choice !== 'auto') {
        return `About ${expectedOnParSeventyTwo(SG_BASELINES_V1[choice].tables)} shots on a par 72.`;
    }
    if (handicapIndex === null) {
        return `No handicap on your profile yet, so this uses the ${cohortLabel('hcp12')} reference.`;
    }
    return `Your ${formatHandicap(handicapIndex)} handicap puts you on the ${cohortLabel(
        cohortForHandicap(handicapIndex),
    )} reference.`;
}

/**
 * A handicap index as it is written everywhere else in the app: one decimal,
 * and a plus handicap reads `+2.0`, not `-2.0`.
 */
export function formatHandicap(index: number): string {
    return index < 0 ? `+${(-index).toFixed(1)}` : index.toFixed(1);
}

/**
 * The tier a choice resolves to.
 *
 * `auto` defers to the handicap, and a missing handicap — no index on the
 * profile, or a profile that has not loaded yet — resolves to `hcp12` through
 * `cohortForHandicap`. A screen drawn before the profile arrives therefore
 * shows today's baseline and then settles, rather than showing nothing.
 */
export function resolveCohort(choice: SgBaselineChoice, handicapIndex: number | null): SgCohort {
    return choice === 'auto' ? cohortForHandicap(handicapIndex) : choice;
}

/**
 * Everything the info sheet needs to say WHICH tier is in force and WHY: the
 * resolved cohort, the setting that produced it, and the handicap the auto mode
 * matched. Carried as one value so a caller cannot pass a cohort from one
 * reading and a handicap from another.
 */
export interface SgBaselineInfo {
    readonly cohort: SgCohort;
    readonly choice: SgBaselineChoice;
    readonly handicapIndex: number | null;
}

/**
 * What a caller with no resolved choice to hand gets: the shipping tier, chosen
 * the way an unresolved `auto` chooses it. Never a claim about the reader —
 * `sgBaselineInfo` is what a screen with a service behind it passes.
 */
export const DEFAULT_SG_BASELINE_INFO: SgBaselineInfo = Object.freeze({
    cohort: cohortForHandicap(null),
    choice: FALLBACK_SG_CHOICE,
    handicapIndex: null,
});

/** Resolve a choice + handicap into the sheet's whole story about the tier. */
export function sgBaselineInfo(
    choice: SgBaselineChoice,
    handicapIndex: number | null,
): SgBaselineInfo {
    return { cohort: resolveCohort(choice, handicapIndex), choice, handicapIndex };
}

function isChoice(raw: string): raw is SgBaselineChoice {
    return (SG_BASELINE_CHOICES as readonly string[]).includes(raw);
}

// Stored as the bare choice string under the SAME key iOS uses, the way
// `tapscore.stats.window.v1` is. An unknown value — a tier renamed, a key
// written by a newer build — decodes to `auto`, which is the setting that
// cannot be wrong for the reader.
const store = deviceStore<SgBaselineChoice>('tapscore.stats.sgBaseline.v1', {
    decode: (raw) => (isChoice(raw) ? raw : FALLBACK_SG_CHOICE),
    encode: (choice) => choice,
    empty: FALLBACK_SG_CHOICE,
});

/** Read the saved baseline choice; `auto` when absent or unreadable. */
export function loadSgChoice(storage: DeviceStorage | null = defaultStorage()): SgBaselineChoice {
    return store.read(storage);
}

/** Persist the baseline choice. A storage failure is swallowed (best-effort). */
export function saveSgChoice(
    choice: SgBaselineChoice,
    storage: DeviceStorage | null = defaultStorage(),
): void {
    store.write(choice, storage);
}

File: ios/TapScore/Features/Stats/SgBaseline.swift
import Foundation

/// Which reference the strokes-gained rows are measured against, as the READER
/// chose it — and the one place that turns that choice into a cohort.
///
/// Twin of the web's `sg-baseline` module: the same five stored strings, the same
/// key, the same resolution rule. A device that has never been touched stores
/// nothing, decodes `.auto`, and lands on the tier the player's handicap implies
/// — `hcp12` when there is no handicap and no session, which is exactly the
/// baseline the app shipped with.

// MARK: - The choice

/// `auto` plus the four tiers. Raw values are the stored strings, identical on
/// both clients.
enum SgBaselineChoice: String, CaseIterable, Sendable {
    /// Follow the signed-in player's handicap index.
    case auto
    case scratch
    case hcp5
    case hcp12
    case hcp20

    /// The tier this choice pins, or nil for `auto` — the one case with no
    /// matching `SgCohort` raw value, which is what makes this lookup total.
    var cohort: SgCohort? { SgCohort(rawValue: rawValue) }

    /// THE resolution rule, and the only one. Both stats surfaces and the ⓘ
    /// sheet go through it, so they cannot disagree about which table a row was
    /// weighed against.
    func resolved(handicapIndex: Double?) -> SgCohort {
        cohort ?? SgCohort.forHandicap(handicapIndex)
    }
}

// MARK: - Words

extension SgCohort {
    /// The tier in the reader's words. Identical strings on both clients.
    var title: String {
        switch self {
        case .scratch: return "Scratch"
        case .hcp5: return "5 handicap"
        case .hcp12: return "12 handicap"
        case .hcp20: return "20+ handicap"
        }
    }

    /// What this tier is expected to shoot on a par 72 — four par 3s, ten par 4s
    /// and four par 5s off its own hole table. Derived, never typed in: a tier
    /// whose table is recalibrated cannot leave a stale number in its subtitle.
    var expectedOnParSeventyTwo: Double {
        let tables = SgBaselines.bundle(for: self).tables
        return 4 * (tables.eHole[3] ?? 0) + 10 * (tables.eHole[4] ?? 0)
            + 4 * (tables.eHole[5] ?? 0)
    }
}

/// The picker's words, in one place so the control and its tests read the same
/// strings. Words over symbols (`docs/design-guidelines.md` §4).
enum SgBaselineCopy {
    static let pickerLabel = "Compared to"
    static let pickerTitle = "Compared to"
    static let autoTitle = "Match my handicap"

    static func rowTitle(_ choice: SgBaselineChoice) -> String {
        choice.cohort?.title ?? autoTitle
    }

    /// The row's explanation — the OPTION's meaning, with the reader's own
    /// handicap where it decides something.
    static func rowSubtitle(_ choice: SgBaselineChoice, handicapIndex: Double?) -> String {
        guard let cohort = choice.cohort else {
            guard let handicapIndex else {
                return
                    "No handicap on your profile yet, so this uses the \(SgCohort.hcp12.title) reference."
            }
            let matched = SgCohort.forHandicap(handicapIndex)
            return
                "Your \(ProfileFormat.index(handicapIndex)) handicap puts you on the \(matched.title) reference."
        }
        return "About \(shots(cohort)) shots on a par 72."
    }

    /// The collapsed field's qualification: which tier the choice actually
    /// resolved to, when the field's own title does not already say it.
    static func fieldMarker(_ context: SgBaselineContext) -> String? {
        guard context.choice == .auto else { return nil }
        return context.cohort.title
    }

    private static func shots(_ cohort: SgCohort) -> String {
        String(Int(cohort.expectedOnParSeventyTwo.rounded()))
    }
}

// MARK: - Persistence

/// The second thing the stats screen remembers between launches: which reference
/// the rows are measured against.
///
/// Same store and same reasoning as `StatsWindowPreference` — `UserDefaults`,
/// not the Keychain and not the server: it is a view preference, it grants
/// nothing, and a device that forgets it falls back to `auto`, which is the
/// answer the app would have computed anyway.
enum SgBaselinePreference {
    /// The web client stores the same string under the same key.
    static let key = "tapscore.stats.sgBaseline.v1"

    /// Follow the handicap. An unreadable or unknown stored value decodes to
    /// this rather than to a tier: a string written by a future version must not
    /// pin a reader to a table this build cannot name.
    static let fallback: SgBaselineChoice = .auto

    static func load(defaults: UserDefaults = .standard) -> SgBaselineChoice {
        guard let raw = defaults.string(forKey: key),
            let choice = SgBaselineChoice(rawValue: raw)
        else { return fallback }
        return choice
    }

    static func save(_ choice: SgBaselineChoice, defaults: UserDefaults = .standard) {
        defaults.set(choice.rawValue, forKey: key)
    }

    /// The stored choice resolved for a caller that has the auth state to hand —
    /// the round screen, the story card and the home card all read the baseline
    /// this way, so none of them can resolve it differently from the dashboard.
    static func context(auth: AuthState, defaults: UserDefaults = .standard) -> SgBaselineContext {
        SgBaselineContext(choice: load(defaults: defaults), handicapIndex: handicapIndex(auth))
    }

    /// The signed-in player's index, or nil for every other state. Negative is a
    /// plus handicap and stays negative — `SgCohort.forHandicap` wants the raw
    /// domain value, not the "+2.4" spelling.
    static func handicapIndex(_ auth: AuthState) -> Double? {
        guard case let .signedIn(player) = auth else { return nil }
        return player.handicapIndex
    }
}

// MARK: - The resolved baseline

/// The baseline in force, and how the reader ended up on it.
///
/// Carried as ONE value rather than as a loose cohort so the ⓘ sheet can say
/// *why* this table — "matched to your 2.0 handicap" and "you chose this" are
/// different sentences, and a bare cohort cannot tell them apart.
struct SgBaselineContext: Equatable, Sendable {
    var choice: SgBaselineChoice
    var handicapIndex: Double?

    init(choice: SgBaselineChoice = .auto, handicapIndex: Double? = nil) {
        self.choice = choice
        self.handicapIndex = handicapIndex
    }

    /// What every surface actually measures against today when nothing has been
    /// resolved: the shipped v1 constants.
    static let fallback = SgBaselineContext()

    var cohort: SgCohort { choice.resolved(handicapIndex: handicapIndex) }

    var bundle: SgBaselineBundle { SgBaselines.bundle(for: cohort) }
}
